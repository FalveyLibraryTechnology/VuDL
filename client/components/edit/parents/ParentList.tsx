import React, { useEffect } from "react";
import { useEditorContext } from "../../../context/EditorContext";
import { useFetchContext } from "../../../context/FetchContext";
import { getParentUrl } from "../../../util/routes";
import Delete from "@mui/icons-material/Delete";

export interface ParentListProps {
    pid: string;
}

const ParentList = ({ pid }: ParentListProps): React.ReactElement => {
    const {
        state: { parentDetailsStorage },
        action: {
            clearPidFromChildListStorage,
            loadParentDetailsIntoStorage,
            removeFromObjectDetailsStorage,
            removeFromParentDetailsStorage,
            setSnackbarState,
        },
    } = useEditorContext();
    const {
        action: { fetchText },
    } = useFetchContext();
    const loaded = Object.prototype.hasOwnProperty.call(parentDetailsStorage, pid);

    useEffect(() => {
        if (!loaded) {
            loadParentDetailsIntoStorage(pid);
        }
    }, [loaded]);

    const showSnackbarMessage = (message: string, severity: string) => {
        setSnackbarState({
            open: true,
            message,
            severity,
        });
    };

    const deleteParent = async (parentPid: string) => {
        if (!confirm("Are you sure you wish to remove this parent?")) {
            return;
        }
        const target = getParentUrl(pid, parentPid);
        let result: string;
        try {
            result = await fetchText(target, { method: "DELETE" });
        } catch (e) {
            result = (e as Error).message ?? "Unexpected error";
        }
        if (result === "ok") {
            // Clear and reload the cached object and its parents, since these have now changed!
            removeFromObjectDetailsStorage(pid);
            removeFromParentDetailsStorage(pid);
            // Clear any cached lists belonging to the parent PID, because the
            // order has potentially changed!
            clearPidFromChildListStorage(parentPid);
            showSnackbarMessage(`Successfully removed ${pid} from ${parentPid}`, "info");
        } else {
            showSnackbarMessage(result, "error");
        }
    };

    const parents = (loaded ? parentDetailsStorage[pid].parents ?? [] : []).map((parent) => {
        let parentChain = "";
        let nextNode = (parent.parents ?? [])[0] ?? null;
        while (nextNode) {
            parentChain = nextNode.title + (parentChain.length ? "/" : "") + parentChain;
            nextNode = (nextNode.parents ?? [])[0] ?? null;
        }
        return (
            <tr key={"parentmodal_" + pid + "_" + parent.pid}>
                <td>
                    <button onClick={() => deleteParent(parent.pid)}>
                        <Delete titleAccess={`Delete parent ${parent.pid}`} />
                    </button>
                </td>
                <td>{parent.pid ?? ""}</td>
                <td>{parent.title ?? "Unknown title"}</td>
                <td>{parentChain}</td>
            </tr>
        );
    });
    return (
        <table border="1">
            <tbody>
                {parents.length > 0 ? (
                    parents
                ) : (
                    <tr key={"parentmodal_" + pid + "_null"}>
                        <td>{loaded ? "No parents defined." : "Loading..."}</td>
                    </tr>
                )}
            </tbody>
        </table>
    );
};

export default ParentList;
