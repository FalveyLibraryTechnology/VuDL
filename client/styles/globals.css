:root,
[color-scheme="light"] {
    --body-bg: #fafbfc;
    --heading-color: #4d4d4d;
    --link-color: #008aed;

    --btn-bg: #485a6c;
    --btn-text: #ffffff;
    --btn-border: #275575;
    --btn-focus-bg: #3498db;

    --btn-primary-bg: #27ae60;
    --btn-primary-text: #ffffff;
    --btn-primary-border: #275575;
    --btn-primary-focus-bg: #155724;

    --btn-danger-bg: red;
    --btn-danger-text: #ffffff;
    --btn-danger-border: #a10705;
    --btn-danger-focus-bg: #a10705;

    --btn-logout-bg: #1abc9c;
    --btn-logout-text: #ffffff;
    --btn-logout-border: #16a085;
    --btn-logout-focus-bg: #1abc9c;

    --btn-toggle-bg: #7f8c8d;
    --btn-toggle-text: #ffffff;
    --btn-toggle-border: #34495e;
    --btn-toggle-focus-bg: #53626f;

    --group-border: #53626f;

    --input-border: #6a737d;
    --input-focus-border: #2188ff;

    --label-hover: #c8e1ff;

    --page-list-bg: #ffffff;
    --page-list-border: #c1c1c1;

    --thumbnail-bg: #e1e1e1;
    --thumbnail-focus-bg: #51aded;
    --thumbnail-number-bg: #ffffff;
    --thumbnail-number-text: #000000;
    --thumbnail-preview-border: #000000;

    --tree-item-focus-bg: #000000;
    --tree-item-focus-text: #ffffff;

    --shadow-gray: #c8e1ff;
}

[color-scheme="dark"] {
}
