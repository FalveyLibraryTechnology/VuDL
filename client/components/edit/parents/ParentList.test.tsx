import React from "react";
import { describe, beforeEach, expect, it, jest } from "@jest/globals";
import { shallow, mount } from "enzyme";
import toJson from "enzyme-to-json";
import ParentList from "./ParentList";

const mockUseEditorContext = jest.fn();
jest.mock("../../../context/EditorContext", () => ({
    useEditorContext: () => {
        return mockUseEditorContext();
    },
}));
const mockUseFetchContext = jest.fn();
jest.mock("../../../context/FetchContext", () => ({
    useFetchContext: () => {
        return mockUseFetchContext();
    },
}));

describe("ParentList", () => {
    let editorValues;
    let fetchValues;
    let pid: string;
    beforeEach(() => {
        pid = "foo:123";
        editorValues = {
            state: {
                parentDetailsStorage: {
                    "foo:123": {
                        parents: [
                            {
                                pid: "foo:122",
                                title: "Parent",
                                parents: [
                                    {
                                        pid: "foo:121",
                                        title: "Grandparent",
                                        parents: [
                                            {
                                                pid: "foo:120",
                                                title: "Great-grandparent",
                                                parents: [],
                                            },
                                        ],
                                    },
                                ],
                            },
                        ],
                    },
                },
            },
            action: {
                clearPidFromChildListStorage: jest.fn(),
                loadParentDetailsIntoStorage: jest.fn(),
                removeFromObjectDetailsStorage: jest.fn(),
                removeFromParentDetailsStorage: jest.fn(),
                setSnackbarState: jest.fn(),
            },
        };
        fetchValues = {
            action: {
                fetchText: jest.fn(),
            }
        };
        mockUseEditorContext.mockReturnValue(editorValues);
        mockUseFetchContext.mockReturnValue(fetchValues);
    });

    it("triggers a data load if necessary", () => {
        editorValues.state.parentDetailsStorage = {};
        const wrapper = mount(<ParentList pid={pid} />);
        expect(editorValues.action.loadParentDetailsIntoStorage).toHaveBeenCalledWith(pid);
    });

    it("renders an empty list correctly", () => {
        editorValues.state.parentDetailsStorage = {
            "foo:123": {
                parents: [],
            }
        };
        const wrapper = shallow(<ParentList pid={pid} />);
        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it("renders a populated parent list correctly", () => {
        const wrapper = shallow(<ParentList pid={pid} />);
        expect(toJson(wrapper)).toMatchSnapshot();
    });
});
