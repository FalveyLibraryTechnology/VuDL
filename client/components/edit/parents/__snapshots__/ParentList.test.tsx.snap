// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ParentList renders a populated parent list correctly 1`] = `
<table
  border="1"
>
  <tbody>
    <tr
      key="parentlist_foo:123_foo:122"
    >
      <td>
        <button
          onClick={[Function]}
        >
          <Memo(ForwardRef(DeleteIcon))
            titleAccess="Delete parent foo:122"
          />
        </button>
      </td>
      <td>
        foo:122
      </td>
      <td>
        Parent
      </td>
      <td>
        Great-grandparent/Grandparent
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`ParentList renders an empty list correctly 1`] = `
<table
  border="1"
>
  <tbody>
    <tr
      key="parentlist_foo:123_null"
    >
      <td>
        No parents defined.
      </td>
    </tr>
  </tbody>
</table>
`;
